### ISSUE TITLE
Allow DuckDB to query a JSON file the same way it can query a CSV.

### ISSUE THREAD

**Issue by lloydtabb** [duckdb/duckdb#3774]
It would be really awesome if you could treat a file containing a JSON array or lines containing single JSON objects as a table and have DuckDB automatically determine the schema.

**Comment by alanpaulkwan** [comment]
i affirm the need of this. apache drill is great for this reason

**Comment by hannes** [comment]
What already works with the [JSON extension](https://duckdb.org/docs/extensions/json) is something like the following, using the second example from https://jsonlines.org/examples/:

```JSON
{"name": "Gilbert", "wins": [["straight", "7♣"], ["one pair", "10♥"]]}
{"name": "Alexa", "wins": [["two pair", "4♠"], ["two pair", "9♠"]]}
{"name": "May", "wins": []}
{"name": "Deloise", "wins": [["three of a kind", "5♣"]]}
```

You can abuse the CSV reader with a non-existing delimiter and cast the single column string column to JSON afterwards:

```SQL
select json(column0) from read_csv_auto('test.json', delim='\0');
```
```
┌─────────────────────────┐
│ json(column0) -> 'name' │
├─────────────────────────┤
│ "Gilbert"               │
│ "Alexa"                 │
│ "May"                   │
│ "Deloise"               │
└─────────────────────────┘
```

Or 

```SQL
select json(column0)->>'name' from read_csv_auto('test.json', delim='\0');
```
```
┌────────────────────────────┐
│ (json(column0) ->> 'name') │
├────────────────────────────┤
│ Gilbert                    │
│ Alexa                      │
│ May                        │
│ Deloise                    │
└────────────────────────────┘
```

**Comment by lnkuiper** [comment]
If you want to go straight to an efficient columnar storage representation, you could do:
```sql
select json_transform(json, '{"name":"VARCHAR","wins":[["VARCHAR"]]}') from read_json_objects('test.json');
┌───────────────────────────────────────────────────────────────────┐
│ json_transform("json", '{"name":"VARCHAR","wins":[["VARCHAR"]]}') │
├───────────────────────────────────────────────────────────────────┤
│ {'name': Gilbert, 'wins': [[straight, 7♣], [one pair, 10♥]]}      │
│ {'name': Alexa, 'wins': [[two pair, 4♠], [two pair, 9♠]]}         │
│ {'name': May, 'wins': []}                                         │
│ {'name': Deloise, 'wins': [[three of a kind, 5♣]]}                │
└───────────────────────────────────────────────────────────────────┘
```
Note that the output consists of our struct/list types, and is no longer a json string.

Of course, this only works if you know the json schema already. You can get the json schema with:
```sql
select json_structure(json) from read_json_objects('test.json');
┌─────────────────────────────────────────┐
│         json_structure("json")          │
├─────────────────────────────────────────┤
│ {"name":"VARCHAR","wins":[["VARCHAR"]]} │
│ {"name":"VARCHAR","wins":[["VARCHAR"]]} │
│ {"name":"VARCHAR","wins":["NULL"]}      │
│ {"name":"VARCHAR","wins":[["VARCHAR"]]} │
└─────────────────────────────────────────┘
```

**Comment by jakthom** [comment]
(Something that might be noteworthy here)

When using either `read_csv_auto('blah.json', delim='\0')` or `read_json_objects('blah.json')` the json must be single-line. Multi-line json in the file causes pain.

**Single line json:**

```
D select * from read_csv_auto('s3://duckling-test/config.json', delim='\0');
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                  column0                                                                                   │
│                                                                                  varchar                                                                                   │
├────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ { "batch_size": 10000, "tables": [ { "database": "postgres", "schema": "public", "table": "users", "load_type": "incremental", "hash_columns": [], "exclude_columns": []…  │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
D select * from read_json_objects('s3://duckling-test/config.json');
┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                                                    json                                                                                    │
│                                                                                    json                                                                                    │
├────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ { "batch_size": 10000, "tables": [ { "database": "postgres", "schema": "public", "table": "users", "load_type": "incremental", "hash_columns": [], "exclude_columns": []…  │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
D
```

**Multi-line json:**

```
D select json(column0) from read_csv_auto('s3://duckling-test/config.json', delim='\0');
Error: Invalid Input Error: malformed JSON
D select * from read_json_objects('s3://duckling-test/config.json');
┌───────────────────────────────────┐
│               json                │
│               json                │
├───────────────────────────────────┤
│ {                                 │
│   "batch_size": 10000,            │
│   "tables": [                     │
│     {                             │
│       "database": "postgres",     │
│       "schema": "public",         │
│       "table": "users",           │
│       "load_type": "incremental", │
│       "hash_columns": [],         │
│       "exclude_columns": []       │
│     }                             │
│   ]                               │
│ }                                 │
├───────────────────────────────────┤
│              13 rows              │
└───────────────────────────────────┘
D select json->'tables' from read_json_objects('s3://duckling-test/config.json');
Error: Invalid Input Error: malformed JSON
D select json->'batch_size' from read_json_objects('s3://duckling-test/config.json');
Error: Invalid Input Error: malformed JSON
```

**Comment by lnkuiper** [comment]
This has been implemented in DuckDB since 0.7.0
